src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -373,17 +374,31 @@ private static void CheckMembers(NamespaceSymbol @namespace, Dictionary<ReadOnly
                 }
             }
 
-            static bool isFileLocalTypeInSeparateFileFrom(SourceMemberContainerTypeSymbol possibleFileLocalType, SourceMemberContainerTypeSymbol otherSymbol)
+            static bool isFileLocalTypeInSeparateFileFrom(SourceMemberContainerTypeSymbol possibleFileLocalType, NamespaceOrTypeSymbol otherSymbol)
             {
                 if (!possibleFileLocalType.IsFileLocal)
                 {
                     return false;
                 }
 
                 var leftTree = possibleFileLocalType.MergedDeclaration.Declarations[0].Location.SourceTree;
-                if (otherSymbol.MergedDeclaration.NameLocations.Any((loc, leftTree) => (object)loc.SourceTree == leftTree, leftTree))
+                if (otherSymbol is SourceNamedTypeSymbol { MergedDeclaration.NameLocations: var typeNameLocations })
                 {
-                    return false;
+                    if (typeNameLocations.Any((loc, leftTree) => (object)loc.SourceTree == leftTree, leftTree))
```

Comment:
nit: consider simplifying the codeflow a bit: `if (otherSymbol is ...) { return !typeNameLocations.Any(...); } else if (otherSymbol is ...) { return !namespaceNameLocations.Any(...) } throw UnexpectedValue;`